vunit/vhdl/python/src/python_pkg_vhpi.c

```diff
@@ -0,0 +1,335 @@
+  const char* unknown_error = "Unknown error";
+
+  // Use provided error reason or try extracting the reason from the Python
+  // exception
+  if (reason == NULL) {
+    PyObject *ptype, *pvalue, *ptraceback;
+    PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+    if (ptype != NULL) {
+      reason = get_string(pvalue);
+    }
+    PyErr_Restore(ptype, pvalue, ptraceback);
+  }
+
+  // Clean-up Python session first in case vhpi_assert stops the simulation
+  if (cleanup) {
+    python_cleanup(NULL);
+  }
+
+  // Output error message
+  reason = reason == NULL ? unknown_error : reason;
+  if (code_or_expr == NULL) {
+    vhpi_assert(vhpiError, "ERROR %s:\n\n%s\n\n", context, reason);
+  } else {
+    vhpi_assert(vhpiError, "ERROR %s:\n\n%s\n\n%s\n\n", context, code_or_expr,
+                reason);
+  }
+
+  // Stop the simulation if vhpi_assert didn't.
+  vhpi_control(vhpiStop);
+}
+
+static void ffi_error_handler(const char* context, bool cleanup) {
+  vhpiErrorInfoT err;
+
+  // Clean-up Python session first in case vhpi_assert stops the simulation
+  if (cleanup) {
+    python_cleanup(NULL);
+  }
+
+  if (vhpi_check_error(&err)) {
+    vhpi_assert(err.severity, "ERROR %s: \n\n%s (%d): %s\n\n", context,
+                err.file, err.line, err.message);
+
+  } else {
+    vhpi_assert(vhpiError, "ERROR %s\n\n", context);
+  }
+
+  // Stop the simulation if vhpi_assert didn't.
+  vhpi_control(vhpiStop);
+}
+
+PLI_VOID python_setup(const struct vhpiCbDataS* cb_p) {
+  Py_Initialize();
+  if (!Py_IsInitialized()) {
+    ffi_error_handler("Failed to initialize Python", false);
+  }
+
+  PyObject* main_module = PyImport_AddModule("__main__");
+  if (main_module == NULL) {
+    ffi_error_handler("Failed to get the main module", true);
+  }
+
+  globals = PyModule_GetDict(main_module);
+  if (globals == NULL) {
+    ffi_error_handler("Failed to get the global dictionary", true);
+  }
+
+  // globals and locals are the same at the top-level
+  locals = globals;
+
+  register_py_error_handler(py_error_handler);
+  register_ffi_error_handler(ffi_error_handler);
+}
+
+PLI_VOID python_cleanup(const struct vhpiCbDataS* cb_p) {
+  if (locals != NULL) {
+    Py_DECREF(locals);
+  }
+
+  if (Py_FinalizeEx()) {
+    vhpi_assert(vhpiWarning, "WARNING: Failed to finalize Python");
+  }
+}
+
+static const char* get_parameter(const struct vhpiCbDataS* cb_p) {
+  // Get parameter from VHDL function call
+  vhpiHandleT parameter_handle =
+      vhpi_handle_by_index(vhpiParamDecls, cb_p->obj, 0);
+  if (parameter_handle == NULL) {
+    ffi_error_handler("getting VHDL parameter handle", true);
+  }
+
+  vhpiValueT parameter;
+  static char vhdl_parameter_string[MAX_VHDL_PARAMETER_STRING_LENGTH];
+
+  parameter.bufSize = MAX_VHDL_PARAMETER_STRING_LENGTH;
+  parameter.value.str = vhdl_parameter_string;
+  parameter.format = vhpiStrVal;
+
+  if (vhpi_get_value(parameter_handle, &parameter)) {
+    ffi_error_handler("getting VHDL parameter value", true);
+  }
+
```

Comment:
I think this needs a call to `vhpi_release_handle(parameter_handle)` here or a it leaks a handle on each invocation.